logger: Add native backend for native_posix

finikorg opened the pull request:
Add backend to be used in native_posix arch.

finikorg committed:
logger: Add native backend for native_posix

Add backend to be used in native_posix arch.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aescolar commented:
Oh well, you overtook me :)

aescolar commented on subsys/logging/log_backend_native.c:
```diff
@@ -0,0 +1,63 @@
+	u32_t flags = 0;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_UART_SHOW_COLOR)) {
+		flags |= LOG_OUTPUT_FLAG_COLORS;
```
Just a note: In principle this is ok, but we need to check if stdout is fed to a file or console.
When fed to a file we should disable it, otherwise all the ascii escapes will be quite annoying.

finikorg replied:
I am thinking to rename `CONFIG_LOG_BACKEND_UART_SHOW_COLOR` to `CONFIG_LOG_SHOW_COLOR` so it might be used in both backends

nordic-krch commented on subsys/logging/CMakeLists.txt:
```diff
@@ -13,3 +13,8 @@ zephyr_sources_ifdef(
   )
+
+zephyr_sources_ifdef(
+  CONFIG_LOG_BACKEND_NATIVE
```
native of which country? ;) to user who do not care about native posix shortcut 'native' does not tell anything thus imo this should rather be log_backend_native_posix.

nordic-krch commented on subsys/logging/log_backend_native.c:
```diff
@@ -0,0 +1,63 @@
+
+	u32_t flags = 0;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_UART_SHOW_COLOR)) {
```
reuse of uart backend option (CONFIG_LOG_BACKEND_UART_SHOW_COLOR) probable copy-paste error.

nordic-krch commented on subsys/logging/log_backend_native.c:
```diff
@@ -0,0 +1,63 @@
+static struct log_output_ctx ctx = {
+	.func = char_out,
+	.data = &buf,
+	.length = 1,
```
yes, so this is a copy-paste of uart backend which is very simple one (polling) and might be reworked in the future. Here, it may make sense to have the buffer bigger if block operations are faster.

nordic-krch commented on subsys/logging/log_backend_native.c:
```diff
@@ -0,0 +1,63 @@
+}
+
+static void panic(struct log_backend const *const backend)
+{
```
are we sure that standard console will always work in panic (e.g. hardfault, error handler?) or it is possible that it relies on interrupt and cannot be used or need to be reconfigured once panic occurs? UART in polling mode will work in that case so panic mode does nothing.

aescolar replied:
In native_posix yes.

aescolar commented:
@finikorg 
Take a look at this branch:
https://github.com/aescolar/zephyr/tree/new_logger
In kind of works.. There is something very fishy happening in the logger itself, which I would need to debug.. I hope I can take a look at it tomorrow.

aescolar commented:
@nordic-krch : Seems this function does not work when the length of the backend buffer is more than 1:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_output.c#L55
Offset grows by 1 each time that function is called and the logger prints its output into the previous string..

aescolar commented:
@nordic-krch : Please take a look at the hacks I did in the 2 commits in the core logger code in this branch:
https://github.com/aescolar/zephyr/tree/new_logger

aescolar committed:
sample: subys: logging: fix for posix arch

In the POSIX arch, time does not pass inside infinite loops.
=> Add a small delay in each iteration of those loops
when compiled for it.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
subsys: logging: Fix for no CONFIG_LOG_PROCESS_THREAD

When CONFIG_LOG_PROCESS_THREAD was not set the logger
was not initialized.
Register a POST_KERNEL init for that case.

Also added an assert in the logger thread in case there
is no backends, instead of having that thread spinning
forever.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

finikorg commented on subsys/logging/log_output.c:
```diff
@@ -59,17 +62,22 @@ static int out_func(int c, void *ctx)
 
 	va_start(args, fmt);
+#ifdef CONFIG_ARCH_POSIX
+	vfprintf(stdout, fmt, args);
```
Otherwise it does not build because of _prf().

finikorg committed:
logger: Rename SHOW_COLOR and FORMAT_TIMESTAMP options

Make options available for other backends

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aescolar committed:
native_posix: Add control of traces color

For native_posix:
Added command line options to control if traces should have
or not colors.
+
Detect if we are connected to a tty or not, and if we are
set the defaults for that option appropriately

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
subsys: logging: for native_posix use appropriate tracing

For the native_posix backend, use the appropiate tracing
functions and enable color if it should be

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
@finikorg : 
It is unfortunate that we both started working on this at the same time. 
(From your other PR comment, I understood you were blocked and waiting for me to add the backend, so I had started on it.)
After I saw yesterday that this PR was not working and had a DNM I just merged it into my branch.
Please look in this branch
https://github.com/aescolar/zephyr/tree/newlogger2
You can take it straight
That branch seems to work, and from my point of view is ready to merge.
It's up to you if you want to use it or you prefer to go with your own changes.

aescolar committed:
tests: subsys: logging: log_core fix with inplace process

tests/subsys/logging/log_core does not work if inplace processing
is set and the process thread is disabled.
So let's set the valid configuration in its project file

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

finikorg committed:
native_posix: Add logger configuration backend

Enable native logger backend for native_posix board in defconfig.
Set also logger configuration options suitable for native_posix
backend.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>
Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
@aescolar I tried to cherry-pick as much your patches as I can, only last one is missing at the moment. I needed this backend for my PR #9877 .

aescolar commented on subsys/logging/Kconfig:
```diff
@@ -270,18 +270,25 @@ config LOG_BACKEND_UART
-	depends on LOG_BACKEND_UART
+config LOG_BACKEND_NATIVE_POSIX
+	bool "Enable native backend"
+	default n
```
Having here the `depends on ARCH_POSIX` would be nice

aescolar commented on subsys/logging/log_backend_native_posix.c:
```diff
@@ -0,0 +1,93 @@
+static struct log_output_ctx ctx = {
+	.func = char_out,
+	.data = buf,
+	.length = 1,
```
I think it is worth while having the comment about why this is just 1 (by now), as otherwise it is difficult to explain why `buf` is 256 char long

aescolar commented on subsys/logging/log_backend_native_posix.c:
```diff
@@ -0,0 +1,93 @@
+	u32_t flags = 0;
+
+	if (IS_ENABLED(CONFIG_LOG_BACKEND_SHOW_COLOR)) {
+		flags |= LOG_OUTPUT_FLAG_COLORS;
```
Could you add this flag only `if (posix_trace_over_tty(0))`?
That way users will still be able to disable it on runtime, and it will be disabled when piping to a file or when run from sanitycheck.

aescolar commented on subsys/logging/log_backend_native_posix.c:
```diff
@@ -0,0 +1,93 @@
+}
+
+static void panic(struct log_backend const *const backend)
+{
```
The comment about why panic does not need to do anything would help other people understand why this can just be stub.
(for ex. @nordic-krch just asked that before)

aescolar commented on subsys/logging/log_core.c:
```diff
@@ -25,6 +26,13 @@ const struct log_backend *uart_backend = &log_backend_uart;
+LOG_BACKEND_DEFINE(log_backend_native, log_backend_native_api);
+const struct log_backend *native_backend = &log_backend_native;
+#else
+const struct log_backend *native_backend;
```
This may be a bit minor:
Each of this is an unused pointer when its backend is not configured in. I would think a `#define native_backend NULL` would be a better choice.

aescolar reviewed (dismissed):
Just a few minor things. 
The only reason for the reject is the color comment.

aescolar commented:
About the 2 failures in CI, they are both due to the new assert that checks that there must be a backend:
https://github.com/zephyrproject-rtos/zephyr/pull/9968/commits/fb0e038c75565b40550bcdf495cd53570a581b0f
That check ensures that the logger thread does not go into an infinite loop while doing nothing.

Both failures are the same, and for xtensa:
```
***** Booting Zephyr OS zephyr-v1.13.0-113-g79cdad94d *****
ASSERTION FAIL [log_backend_count_get() > 0] @ /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/logging/log_core.c:504:
```
It seems these tests were passing before, because they do not actually require a backend. Now they do not because as soon as the assert is triggered that logger thread is killed.

@finikorg : I guess you could remove that assert as it is not the point of this PR to fix issues in the logger itself. But @nordic-krch : maybe you want to consider having that assert, and looking into the issue for this platform(?)

finikorg committed:
logger: Add init() api and simplify init and activate

Simplify logger initialization process using already existing loops.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
logger: Move native_posix_backend definition

Move definition to appropriate place removing ugly defines.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
logger: Simplify UART backend initialization

Move definition to sources of UART backend.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aescolar reviewed (commented):
Ok by me.
Minor: You may want to change the commit messages for
* "native_posix: Add logger configuration"
(The description should probably say what the commit does. You can just add my signoff to it instead of mentioning in the message you got the change from my branch)
* "logger: Remove assert() in thread start" Could probably use a longer explanation of why that is being done (OR you can just squash that commit together into the commit where I added it, and save yourself from explaining why you remove it)

dbkinder reviewed (approved):
+1 for doc changes

aescolar commented:
@finikorg Note that the changes to the logger merged in master are conflicting with this changes.

finikorg commented:
@aescolar rebased

aescolar commented:
@nordic-krch : are you happy with this new backend and with the piggybacked bugfixes in this PR?

aescolar commented on subsys/logging/log_core.c:
```diff
@@ -18,13 +17,6 @@
 #define CONFIG_LOG_PRINTK_MAX_STRING_LENGTH 1
 #endif
 
-#ifdef CONFIG_LOG_BACKEND_UART
```
+1 for not having the different backends code in the common logger files.

aescolar reviewed (approved):
+2 to the backend
+1 for the refactoring in the logger code and bugfixes

nordic-krch commented on subsys/logging/log_core.c:
```diff
@@ -256,14 +248,17 @@ void log_init(void)
+			backend->api->init();
+		}
+
+		log_backend_activate(backend, NULL);
```
There will be use cases where we don't want to enable all backends. Example: log backend over bluetooth where backend will be registered but not activated until connection is established.

aescolar replied:
Is the expectation then, that backends which are meant to be started if configured, be called from here (like the UART one was, but the others be activated from other places?

nordic-krch replied:
yes, i was thinking the same that we might want to extend backend registeration macro to has something autostart flag.

nordic-krch replied:
Lets keep this change as is and i will try to address that in separate PR.

aescolar replied:
> Lets keep this change as is and i will try to address that in separate PR.

That would be best yes.

>  extend backend registeration macro to has something autostart flag.

Note that would use memory, while keeping that information implicitly in the order would not.

finikorg replied:
I think we might handle some special cases in api->init()

aescolar commented on samples/subsys/logging/logger/src/main.c:
```diff
@@ -171,6 +171,9 @@ static void performance_showcase(void)
 
 	while (start_timestamp == timestamp_get()) {
+#if (CONFIG_ARCH_POSIX)
+		k_busy_wait(100);
```
This app was hanging in CI when run in native_posix. 
This is due to the way the posix arch works: *You can imagine the code executes in a simulated CPU which runs at an infinitely fast clock: No time passes while the CPU is running.*
https://docs.zephyrproject.org/latest/boards/posix/native_posix/doc/board.html#important-limitations

aescolar commented:
> Seems this function does not work when the length of the backend buffer is more than 1

@nordic-krch : Was this fixed? At least if I try now to increase the registered output size things seem to work ok.

aescolar commented:
@nordic-krch : Is this PR good enough to get your approval?

nordic-krch commented:
@aescolar 
> Was this fixed?

yes

>  Is this PR good enough to get your approval?

yes

aescolar commented:
@carlescufi / @nashif Could either of you do a quick review, and if you are happy with it merge?

carlescufi commented:
@aescolar I'm fine with this but would like @nordic-krch to approve before we merge.

aescolar commented:
He has on the comment https://github.com/zephyrproject-rtos/zephyr/pull/9968#issuecomment-422361401

aescolar commented:
@carlescufi : He has on the comment just above https://github.com/zephyrproject-rtos/zephyr/pull/9968#issuecomment-422361401
